File: packages-gen/dart/libs/xt-ui/lib/ui/view/backend.dart
import 'package:xtalk_substrate/view-catalog.dart' as catalog;


native_registry() {
  return <dynamic, dynamic>{
    "ui/description":<dynamic, dynamic>{"type":"WText","value_prop":"value"},
    "ui/column":<dynamic, dynamic>{"type":"WDiv","layout":"column"},
    "ui/row":<dynamic, dynamic>{"type":"WDiv","layout":"row"},
    "ui/textarea":<dynamic, dynamic>{"type":"WInput","input":true},
    "ui/icon":<dynamic, dynamic>{"type":"WIcon","value_prop":"value"},
    "ui/text":<dynamic, dynamic>{"type":"WText","value_prop":"value"},
    "ui/label":<dynamic, dynamic>{"type":"WText","value_prop":"value"},
    "ui/input":<dynamic, dynamic>{"type":"WInput","input":true},
    "ui/spinner":<dynamic, dynamic>{"type":"WText","value_prop":"value"},
    "ui/title":<dynamic, dynamic>{"type":"WText","value_prop":"value","role":"title"},
    "ui/alert":<dynamic, dynamic>{"type":"WDiv","layout":"column"},
    "ui/fragment":<dynamic, dynamic>{"type":"WDiv","layout":"column"},
    "ui/button":<dynamic, dynamic>{"type":"WButton","press":true},
    "ui/image":<dynamic, dynamic>{"type":"WImage"},
    "ui/scroll":<dynamic, dynamic>{"type":"WDiv","layout":"column"}
  };
}

native_entry(registry, component_id) {
  if((() {
    var dart_truthy__53181 = catalog.platform_idp(component_id);
    return (null != dart_truthy__53181) && (false != dart_truthy__53181);
  })()){
    throw "platform view component not portable [wind] - " + component_id;
  }
  return (registry["native"])[component_id];
}

registry(overrides, polyfills) {
  return <dynamic, dynamic>{
    "backend":"wind",
    "native":native_registry(),
    "polyfills":polyfills ?? <dynamic, dynamic>{},
    "overrides":overrides ?? <dynamic, dynamic>{}
  };
}

action_add(runtime, state, event_id, action_desc, value_event) {
  var action_id = "substrate_view_" + (state["next"]).toString();
  state["next"] = (1 + state["next"]);
  state["actions"][action_id] = ((args) {
    var event = ((null != value_event) && (false != value_event)) ? <dynamic, dynamic>{"value":args["_value"]} : (args ?? <dynamic, dynamic>{});
    return runtime["dispatch"](action_desc,event);
  });
  return <dynamic, dynamic>{event_id:<dynamic, dynamic>{"action":action_id}};
}

props(runtime, component_id, input, state, entry) {
  var out = <dynamic, dynamic>{};
  for(var entry_53182 in (input ?? <dynamic, dynamic>{}).entries){
    var key = entry_53182.key;
    var value = entry_53182.value;
    if(key == "class"){
      out["className"] = value;
    }
    else if(key == "aria_label"){
      out["semanticLabel"] = value;
    }
    else if(key == "read_only"){
      out["readOnly"] = value;
    }
    else if(key == "pending"){
      out["isLoading"] = value;
    }
    else if(key == "hidden"){
      null;
    }
    else if(key == "variant"){
      null;
    }
    else if(key == "on_press"){
      xt.lang.common_data.obj_assign(out,action_add(runtime,state,"onTap",value,false));
    }
    else if(key == "on_change"){
      xt.lang.common_data.obj_assign(out,action_add(runtime,state,"onChange",value,true));
    }
    else{
      out[key] = value;
    }
  };
  var variant = input["variant"];
  if(null != variant){
    var classes = catalog.variant_classes(component_id,variant);
    if(null != classes){
      var current = out["className"] ?? "";
      if(0 < current.length){
        current = (current + " ");
      }
      out["className"] = (current + classes);
    }
  }
  var layout = entry["layout"];
  if((null != layout) && (false != layout)){
    var base = (layout == "row") ? "flex flex-row" : "flex flex-col";
    out["className"] = (base + " " + (out["className"] ?? ""));
  }
  var value_prop = entry["value_prop"];
  if((null != value_prop) && (false != value_prop)){
    out["text"] = (input[value_prop] ?? "");
    out.remove(value_prop);
  }
  return out;
}

prepare_native(runtime, component_id, entry, props, children, state) {
  return <dynamic, dynamic>{
    "type":entry["type"],
    "props":props(runtime,component_id,props,state,entry),
    "children":children
  };
}

File: src-lang/dart/ui/view/backend.clj
(ns dart.ui.view.backend
  "Native Wind widget mappings for substrate view component ids.

   Styling is Tailwind class strings only (`class` -> `className`); `variant`
   props lower to the shared class bundles in xt.substrate.view-catalog.
   Platform (`fg/`) ids are figma-only and rejected here."
  (:require [hara.lang :as l]))

(l/script :dart
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]
             [xt.lang.common-string :as xts]
             [xt.substrate.view-catalog :as catalog]]})

(defn.dt native-registry
  []
  (return
   {"ui/fragment"    {"type" "WDiv" "layout" "column"}
    "ui/row"         {"type" "WDiv" "layout" "row"}
    "ui/column"      {"type" "WDiv" "layout" "column"}
    "ui/scroll"      {"type" "WDiv" "layout" "column"}
    "ui/text"        {"type" "WText" "value_prop" "value"}
    "ui/title"       {"type" "WText" "value_prop" "value" "role" "title"}
    "ui/label"       {"type" "WText" "value_prop" "value"}
    "ui/description" {"type" "WText" "value_prop" "value"}
    "ui/icon"        {"type" "WIcon" "value_prop" "value"}
    "ui/input"       {"type" "WInput" "input" true}
    "ui/textarea"    {"type" "WInput" "input" true}
    "ui/button"      {"type" "WButton" "press" true}
    "ui/alert"       {"type" "WDiv" "layout" "column"}
    "ui/spinner"     {"type" "WText" "value_prop" "value"}
    "ui/image"       {"type" "WImage"}}))

(defn.dt native-entry
  "resolves a component id to a native entry, rejecting `fg/` platform ids"
  [registry component-id]
  (when (catalog/platform-id? component-id)
    (xt/x:err (xt/x:cat "platform view component not portable [wind] - "
                        component-id)))
  (return (xt/x:get-key (xt/x:get-key registry "native") component-id)))

(defn.dt registry
  [overrides polyfills]
  (return {"backend" "wind"
           "native" (-/native-registry)
           "polyfills" (or polyfills {})
           "overrides" (or overrides {})}))

(defn.dt action-add
  [runtime state event-id action-desc value-event]
  (var action-id (xt/x:cat "substrate_view_"
                           (xt/x:to-string (xt/x:get-key state "next"))))
  (xt/x:set-key state "next" (+ 1 (xt/x:get-key state "next")))
  (xt/x:set-key
   (xt/x:get-key state "actions") action-id
   (fn [args]
     (var event (:? value-event
                    {"value" (xt/x:get-key args "_value")}
                    (or args {})))
     (return ((xt/x:get-key runtime "dispatch") action-desc event))))
  (return {event-id {"action" action-id}}))

(defn.dt props
  "normalizes portable props to Wind props; styling is classes plus the
   shared variant class bundles"
  [runtime component-id input state entry]
  (var out {})
  (xt/for:object [[key value] (or input {})]
    (cond (== key "class") (xt/x:set-key out "className" value)
          (== key "aria_label") (xt/x:set-key out "semanticLabel" value)
          (== key "read_only") (xt/x:set-key out "readOnly" value)
          (== key "pending") (xt/x:set-key out "isLoading" value)
          (== key "hidden") nil
          (== key "variant") nil
          (== key "on_press")
          (xtd/obj-assign out (-/action-add runtime state "onTap" value false))
          (== key "on_change")
          (xtd/obj-assign out (-/action-add runtime state "onChange" value true))
          :else (xt/x:set-key out key value)))
  (var variant (xt/x:get-key input "variant"))
  (when (xt/x:not-nil? variant)
    (var classes (catalog/variant-classes component-id variant))
    (when (xt/x:not-nil? classes)
      (var current (or (xt/x:get-key out "className") ""))
      (when (< 0 (xt/x:str-len current))
        (:= current (xt/x:cat current " ")))
      (xt/x:set-key out "className" (xt/x:cat current classes))))
  (var layout (xt/x:get-key entry "layout"))
  (when layout
    (var base (:? (== layout "row") "flex flex-row" "flex flex-col"))
    (xt/x:set-key out "className"
                  (xt/x:cat base " " (or (xt/x:get-key out "className") ""))))
  (var value-prop (xt/x:get-key entry "value_prop"))
  (when value-prop
    (xt/x:set-key out "text" (or (xt/x:get-key input value-prop) ""))
    (xt/x:del-key out value-prop))
  (return out))

(defn.dt prepare-native
  [runtime component-id entry input children state]
  (return {"type" (xt/x:get-key entry "type")
           "props" (-/props runtime component-id input state entry)
           "children" children}))
